### PR TITLE
Hyper-V VM Enumeration Module

### PR DESCRIPTION
**Pull request by gwillcox-r7**
This PR adds a very simple auxiliary module which will see if a target is a Hyper-V host. If it is, it will attempt to gather information about all Hyper-V VMs that exist on the target using the `Get-VM` PowerShell command, and it will then save these results into the loot for later use.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] Get a Meterpreter shell on a Hyper-V host.
- [ ] `use use post/windows/gather/enum_hyperv_vms`
- [ ] `set SESSION <session ID>`
- [ ] `run`
- [ ] **Verify** that the exploit returns an error if the current user does not have permissions to run the `Get-VM` PowerShell commandlet.
- [ ] **Verify** that if the current user does have permissions to run the `Get-VM`  commandlet, that details about all of the VMs that exist on the target are returned and saved into a loot file.
- [ ] **Document** anything that is not working correctly.

### CODE REVIEW

File: modules/post/windows/gather/enum_hyperv_vms.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    end
+    results = psh_exec('Get-VM')
+    if results =~ /is not recognized as the name of a cmdlet/
+      print_error('The target is not a Hyper-V host')
+    elsif results =~ /do not have the required permission/
+      print_error('You need to be running as an elevated admin or a user of the Hyper-V Administrators group to run this module')
+      return
+    end
+    vprint_status(results)
+    filtered_result = results.match(/^Name(.+\r\n){1,2000}/) # If your running more than 2000 VMs on a single host, you have my sincerest sympathy.
```
Comment (cdelafuente-r7) [inline modules/post/windows/gather/enum_hyperv_vms.rb:41]: When captured result is not used, it is preferred to use a non-capture group with`(?:`.
```suggestion
    filtered_result = results.match(/^Name(?:.+\r\n){1,2000}/) # If your running more than 2000 VMs on a single host, you have my sincerest sympathy.
```

**Review by cdelafuente-r7 — COMMENTED** [review]
@gwillcox-r7 , thanks for this module. It looks good to me. I just left a couple of minor comments.

**Comment by cdelafuente-r7** [comment]
I tested against Windows Server 2019 Version 1809 with Hyper-V and verified the information related to VMs on the target are returned and saved into a loot file. Also, I verified the exploit returns an error if the current user does not have permissions to run the `Get-VM` PowerShell commandlet.

- output with a privileged user (Administrator with verbose mode):
```
msf6 post(windows/gather/enum_hyperv_vms) > run

[+] Compressed size: 800
[+] Final command JgAoAFsAcwBjAHIAaQBwAHQAYgBsAG8AYwBrAF0AOgA6AGMAcgBlAGEAdABlACgAKABOAGUAdwAtAE8AYgBqAGUAYwB0ACAAUwB5AHMAdABlAG0ALgBJAE8ALgBTAHQAcgBlAGEAbQBSAGUAYQBkAGUAcgAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEcAegBpAHAAUwB0AHIAZQBhAG0AKAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAE0AZQBtAG8AcgB5AFMAdAByAGUAYQBtACgALABbAFMAeQBzAHQAZQBtAC4AQwBvAG4AdgBlAHIAdABdADoAOgBGAHIAbwBtAEIAYQBzAGUANgA0AFMAdAByAGkAbgBnACgAJwBIADQAcwBJAEEASwBIAGoAWQBWADgAQwBBADMATgBQAEwAZABFAE4AOAB3AFUAQQB6AFUAWQBKAHcAZwBZAEEAQQBBAEEAPQAnACkAKQApACwAWwBTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgBNAG8AZABlAF0AOgA6AEQAZQBjAG8AbQBwAHIAZQBzAHMAKQApACkALgBSAGUAYQBkAFQAbwBFAG4AZAAoACkAKQApAGUAYwBoAG8AIAAnAHMAVQBOAFQAZgBGAEIAcgAnADsA
[+] EXECUTING:
powershell.exe -EncodedCommand JgAoAFsAcwBjAHIAaQBwAHQAYgBsAG8AYwBrAF0AOgA6AGMAcgBlAGEAdABlACgAKABOAGUAdwAtAE8AYgBqAGUAYwB0ACAAUwB5AHMAdABlAG0ALgBJAE8ALgBTAHQAcgBlAGEAbQBSAGUAYQBkAGUAcgAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEcAegBpAHAAUwB0AHIAZQBhAG0AKAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAE0AZQBtAG8AcgB5AFMAdAByAGUAYQBtACgALABbAFMAeQBzAHQAZQBtAC4AQwBvAG4AdgBlAHIAdABdADoAOgBGAHIAbwBtAEIAYQBzAGUANgA0AFMAdAByAGkAbgBnACgAJwBIADQAcwBJAEEASwBIAGoAWQBWADgAQwBBADMATgBQAEwAZABFAE4AOAB3AFUAQQB6AFUAWQBKAHcAZwBZAEEAQQBBAEEAPQAnACkAKQApACwAWwBTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgBNAG8AZABlAF0AOgA6AEQAZQBjAG8AbQBwAHIAZQBzAHMAKQApACkALgBSAGUAYQBkAFQAbwBFAG4AZAAoACkAKQApAGUAYwBoAG8AIAAnAHMAVQBOAFQAZgBGAEIAcgAnADsA -InputFormat None
[+] Cleaning up 1836
[+] EXECUTING:
powershell.exe -EncodedCommand JgAoAFsAcwBjAHIAaQBwAHQAYgBsAG8AYwBrAF0AOgA6AGMAcgBlAGEAdABlACgAKABOAGUAdwAtAE8AYgBqAGUAYwB0ACAAUwB5AHMAdABlAG0ALgBJAE8ALgBTAHQAcgBlAGEAbQBSAGUAYQBkAGUAcgAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEcAegBpAHAAUwB0AHIAZQBhAG0AKAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAE0AZQBtAG8AcgB5AFMAdAByAGUAYQBtACgALABbAFMAeQBzAHQAZQBtAC4AQwBvAG4AdgBlAHIAdABdADoAOgBGAHIAbwBtAEIAYQBzAGUANgA0AFMAdAByAGkAbgBnACgAJwBIADQAcwBJAEEASwBYAGoAWQBWADgAQwBBADQAdAAyAHoAUwB2AEwATABNAHIAUAB5ADAAMwBOAEsANABtADEAcwBuAEoAUABMAFUARQBTAEMARQBzAHMAeQBrAHgATQB5AGsAawB0ADEAbABBAFAATABVADQAdABVAHQAZgBVAHkAMAA2AHQATABLADQASgBUAHMAMQBKAFQAUwA3AFIARABTADQAcAB5AHMAeABMAFYAdwBoAE0AQwB2AFMAcwBMAEMANgBMAHIARgBHAHQAagBrAFkAMQBMAEIAaQByAFkAUgBvAHEAOABUAG8AcQBlAGEAVQA1AE8AVABwAFEAUQAyAHMAQgA5AGwAQQAzAHQASQBRAEEAQQBBAEEAPQAnACkAKQApACwAWwBTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgBNAG8AZABlAF0AOgA6AEQAZQBjAG8AbQBwAHIAZQBzAHMAKQApACkALgBSAGUAYQBkAFQAbwBFAG4AZAAoACkAKQApAGUAYwBoAG8AIAAnAHMAVQBOAFQAZgBGAEIAcgAnADsA -InputFormat None
[*] #< CLIXML

Name    State CPUUsage(%) MemoryAssigned(M) Uptime   Status             Version
----    ----- ----------- ----------------- ------   ------             -------
Win7 #1 Off   0           0                 00:00:00 Operating normally 8.2


<Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj><Obj S="progress" RefId="1"><TNRef RefId="0" /><MS><I64 N="SourceId">2</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>
[*] Name    State CPUUsage(%) MemoryAssigned(M) Uptime   Status             Version
----    ----- ----------- ----------------- ------   ------             -------
Win7 #1 Off   0           0                 00:00:00 Operating normally 8.2

[+] Stored loot at /home/msfuser/.msf4/loot/20200916120630_default_172.31.22.124_host.hyperv_vms_140479.txt
[*] Post module execution completed
```
- output with a non-privileged user (with verbose mode):
```
msf6 post(windows/gather/enum_hyperv_vms) > run

[+] Compressed size: 800
[+] Final command JgAoAFsAcwBjAHIAaQBwAHQAYgBsAG8AYwBrAF0AOgA6AGMAcgBlAGEAdABlACgAKABOAGUAdwAtAE8AYgBqAGUAYwB0ACAAUwB5AHMAdABlAG0ALgBJAE8ALgBTAHQAcgBlAGEAbQBSAGUAYQBkAGUAcgAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEcAegBpAHAAUwB0AHIAZQBhAG0AKAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAE0AZQBtAG8AcgB5AFMAdAByAGUAYQBtACgALABbAFMAeQBzAHQAZQBtAC4AQwBvAG4AdgBlAHIAdABdADoAOgBGAHIAbwBtAEIAYQBzAGUANgA0AFMAdAByAGkAbgBnACgAJwBIADQAcwBJAEEATQBmAHAAWQBWADgAQwBBADMATgBQAEwAZABFAE4AOAB3AFUAQQB6AFUAWQBKAHcAZwBZAEEAQQBBAEEAPQAnACkAKQApACwAWwBTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgBNAG8AZABlAF0AOgA6AEQAZQBjAG8AbQBwAHIAZQBzAHMAKQApACkALgBSAGUAYQBkAFQAbwBFAG4AZAAoACkAKQApAGUAYwBoAG8AIAAnAHQAYgBwAEEAcQBpAGUASgAnADsA
[+] EXECUTING:
powershell.exe -EncodedCommand JgAoAFsAcwBjAHIAaQBwAHQAYgBsAG8AYwBrAF0AOgA6AGMAcgBlAGEAdABlACgAKABOAGUAdwAtAE8AYgBqAGUAYwB0ACAAUwB5AHMAdABlAG0ALgBJAE8ALgBTAHQAcgBlAGEAbQBSAGUAYQBkAGUAcgAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEcAegBpAHAAUwB0AHIAZQBhAG0AKAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAE0AZQBtAG8AcgB5AFMAdAByAGUAYQBtACgALABbAFMAeQBzAHQAZQBtAC4AQwBvAG4AdgBlAHIAdABdADoAOgBGAHIAbwBtAEIAYQBzAGUANgA0AFMAdAByAGkAbgBnACgAJwBIADQAcwBJAEEATQBmAHAAWQBWADgAQwBBADMATgBQAEwAZABFAE4AOAB3AFUAQQB6AFUAWQBKAHcAZwBZAEEAQQBBAEEAPQAnACkAKQApACwAWwBTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgBNAG8AZABlAF0AOgA6AEQAZQBjAG8AbQBwAHIAZQBzAHMAKQApACkALgBSAGUAYQBkAFQAbwBFAG4AZAAoACkAKQApAGUAYwBoAG8AIAAnAHQAYgBwAEEAcQBpAGUASgAnADsA -InputFormat None
[+] Cleaning up 3608
[+] EXECUTING:
powershell.exe -EncodedCommand JgAoAFsAcwBjAHIAaQBwAHQAYgBsAG8AYwBrAF0AOgA6AGMAcgBlAGEAdABlACgAKABOAGUAdwAtAE8AYgBqAGUAYwB0ACAAUwB5AHMAdABlAG0ALgBJAE8ALgBTAHQAcgBlAGEAbQBSAGUAYQBkAGUAcgAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEcAegBpAHAAUwB0AHIAZQBhAG0AKAAoAE4AZQB3AC0ATwBiAGoAZQBjAHQAIABTAHkAcwB0AGUAbQAuAEkATwAuAE0AZQBtAG8AcgB5AFMAdAByAGUAYQBtACgALABbAFMAeQBzAHQAZQBtAC4AQwBvAG4AdgBlAHIAdABdADoAOgBGAHIAbwBtAEIAYQBzAGUANgA0AFMAdAByAGkAbgBnACgAJwBIADQAcwBJAEEATQB2AHAAWQBWADgAQwBBADQAdAAyAHoAUwB2AEwATABNAHIAUAB5ADAAMwBOAEsANABtADEAcwBuAEoAUABMAFUARQBTAEMARQBzAHMAeQBrAHgATQB5AGsAawB0ADEAbABBAFAATABVADQAdABVAHQAZgBVAHkAMAA2AHQATABLADQASgBUAHMAMQBKAFQAUwA3AFIARABTADQAcAB5AHMAeABMAFYALwBCAHoAagBBAHoAMAB5AGMAeAAwAHEAbABHAHQAagBrAFkAMQBMAEIAaQByAFkAUgBvAHEAOABUAG8AcQBlAGEAVQA1AE8AVABwAFEAUQAyAHMAQgBLADUAbQBXAHEAbwBRAEEAQQBBAEEAPQAnACkAKQApACwAWwBTAHkAcwB0AGUAbQAuAEkATwAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgAuAEMAbwBtAHAAcgBlAHMAcwBpAG8AbgBNAG8AZABlAF0AOgA6AEQAZQBjAG8AbQBwAHIAZQBzAHMAKQApACkALgBSAGUAYQBkAFQAbwBFAG4AZAAoACkAKQApAGUAYwBoAG8AIAAnAHQAYgBwAEEAcQBpAGUASgAnADsA -InputFormat None
[-] You need to be running as an elevated admin or a user of the Hyper-V Administrators group to run this module
[*] Post module execution completed
```

File: documentation/modules/post/windows/gather/enum_hyperv_vms.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  1. Start `msfconsole`
+  2. Get meterpreter session
+  3. Do: `use post/windows/gather/enum_hyperv_vms`
+  4. Do: `set SESSION <session id>`
+  5. Do: `run`
+  6. If the host has Hyper-V installed, a list of Hyper-V VMs which are on target host will be returned, along with their attributes.
+
+## Options
+
+  - **SESSION** - The session to run the module on.
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply documentation/modules/post/windows/gather/enum_hyperv_vms.md:18]: Woops didn't realize I still had this in here, I'll remove this so long.

File: modules/post/windows/gather/enum_hyperv_vms.rb (hunk 1 above)
Comment (gwillcox-r7) [reply modules/post/windows/gather/enum_hyperv_vms.rb:41]: Yep sounds good, still learning some regex and forgot about this when I wrote this line, but changes make sense 👍

File: documentation/modules/post/windows/gather/enum_hyperv_vms.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply documentation/modules/post/windows/gather/enum_hyperv_vms.md:18]: Going to replace this with a one liner but if you have any ideas for this section feel free to update. @adamgalway-r7 or one of his team members may want to look at updating the RuboCop template as right now if we remove this section, we will get RuboCop failures.

[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply documentation/modules/post/windows/gather/enum_hyperv_vms.md]: Ok changes made, @cdelafuente-r7 feel free to make any changes you feel are needed to correct this if needs be prior to landing.

[the remark this answers is not in the document]
Comment (cdelafuente-r7) [reply documentation/modules/post/windows/gather/enum_hyperv_vms.md]: Thanks! It looks good to me.

**Comment by cdelafuente-r7** [comment]
Thanks @gwillcox-r7! Everything looks good to me and I will land it shortly.

**Comment by cdelafuente-r7** [comment]
# Release Notes
This auxiliary module check if a target is a Hyper-V host and attempt to gather information about all Hyper-V VMs.

**Pull request merged**